### ISSUE TITLE
Ctrl-C doesn't terminate process or clear the line

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/node-pty#7]
From https://github.com/Microsoft/vscode/issues/15249

> - VSCode Version: 1.7.1
> - OS Version: Windows 8.1
> 
> Steps to Reproduce:
> 
> 1. "terminal.integrated.shell.windows": "C:\\Program Files\\Git\\bin\\bash.exe"
> 2. Ctrl-C doesn't work for terminating a process (such as cordova run).

**Comment by Syllinger** [comment]
Also having an issue with Ctrl-C where the current process is not halted, instead the terminal is closed/crashes and all running processes exit in unison.

- VSCode Version: 1.7.1
- OS Version: Windows 10 Home

**Comment by Tyriar** [comment]
@Syllinger If you see that again can you go to Hep > Toggle Developer Tools, open the console and grab a screenshot and/or paste the error?

**Comment by Tyriar** [comment]
This should be fixed in upcoming release.

**Comment by Tyriar** [comment]
Sending `^C` to `gulp watch` sometimes seems to do nothing, however `^C` does work in general. Not sure why atm.

**Comment by melgish** [comment]
Observed similar behavior running live-server.  Also noticed that sometimes CTRL-C works but only after a lengthy timeout.

**Comment by CherryDT** [comment]
As mentioned in the VSCode issue already: This happens to everybody in our company as well. However it also happens to outside VSCode, I have to add. We use ConEmu, so maybe the issue exists in general with the combination of Git Bash, npm/node and non-default console emulators?

This makes me wonder who is really at fault here, it could be very well the MSYS bash...

**Comment by CherryDT** [comment]
In fact it sounds like https://github.com/Maximus5/ConEmu/issues/402 and https://github.com/git-for-windows/git/issues/227

**Comment by vladbalanescu** [comment]
Any updates on this issue guys? I am currently running a continuous Casper script in my bash shell on Windows and I cannot stop the execution other than after finishing the promise chain execution.

**Comment by IanOvenden** [comment]
Hey @vladbalanescu - although not perfect within VSCode, I did find that upgrading to latest version of Node allowed me to terminate most processes :+1:

**Comment by vladbalanescu** [comment]
Hey @IanOvenden which version of node are you using? I am currently on 7.10 and whenever I am running a casperjs script I cannot stop it unless it finished execution and the promise chain finished returning

**Comment by IanOvenden** [comment]
@vladbalanescu - I'm currently running on version 6.11.1, which is the latest stable release.  At the time I was having issues, I was on 4.x.x and upgraded to an early version of 6 - I can't remember exactly which versions I moved from.

**Comment by tateman74** [comment]
Getting the same thing guys with a node or nodemon process.  Dies and leaves the process running.  I'm running Node 8.2.1 so there is no fix with a node upgrade.  This is a Bash issue with Windows 10 only.  I'm going to have to use a different shell now as I can't keep hunting down the node process to kill it.

<img width="455" alt="screen shot 2017-08-01 at 11 07 49 am" src="https://user-images.githubusercontent.com/28845493/28837344-ad1ee806-76a9-11e7-9f01-b7f7cab3b194.png">

**Comment by slavik57** [comment]
Same here...

I have a simple express app that listens on some port...
When I run it with `node server.js` it passes the CTRL+C to it and terminates it
When I run it with `npm start` when the start is `"start": "node server.js"` and press CTRL+C it hangs...

I noticed that when running with default windows command line it doesn't happen...

**Comment by kucukkanat** [comment]
@slavik57 this is very useful information thanks

**Comment by parkovski** [comment]
I’ve been trying to figure this one out and it’s pretty baffling but I’ve got a bunch of info that may help.

The problem is in winpty, src/agent/ConsoleInput.cc around line 400 where it generates a Ctrl C event. Powershell is easy - PSReadLine disables processed mode and responds to the ^C character. All you have to do is WriteConsoleInput a \x03 and it will work. This has to happen unconditionally though, the processed mode check in winpty makes it not work somehow.

The one that’s got me stuck, and I’ve spent hours in windbg on this, is cmd. Its input loop is basically call ReadConsole and check a global flag if a Ctrl C was generated during the call. Then in another thread, it has a control event handler that sets the flag when the system event is triggered. From the disassembly it looks like it *should* happen when you call GenerateConsoleControlEvent, but it doesn’t. Or rather, the event clearly goes through, because I can get a “terminate batch job” prompt, but pressing it again does nothing, and it doesn’t work for the regular prompt either. The system frustratingly does something different when a debugger is active so it’s really hard to track this one down.

To get this to work on top of what I have working (using calling GenerateConsoleCtrlEvent _and_ sending a \x03), all that’s left is to get cmd to recognize the event when it’s in a prompt.

A possible hint is that cmd normally prints two new lines after an actual Ctrl C press, but if you disable processed mode it only prints one. So somehow I think there are two separate checks for this and I’m missing both of them.

**Comment by parkovski** [comment]
After a bit more looking at the disassembly, there clearly are two checks. The first is in `Handler`, the Ctrl-C/Break handler, which is frustratingly really straightforward. The very top of the function essentially says `if (ecx < 2) SetCtrlC();` where `ecx` is the control code - 0 and 1 are Ctrl-C and Ctrl-Break.

The other check is for the actual character `\x03` as it's reading the console buffer into the lexer buffer. There is something going on with a key flag and a data flag that I'm not sure of though, and it's pretty hard to follow because it uses `setjmp/longjmp` and frequently jumps to this big block of code soup called `chkstk` that's just a bunch of random pieces of other functions.

**Comment by sigod** [comment]
> `git --version` outputs: 2.13.3.windows.1

`2.14.2.windows.2` — the same issue.
`2.13.0.windows.1` — the issue isn't present.

**Comment by slavik57** [comment]
@sigod seems like it's time to `git bisect` the problem

**Comment by yayayahahaha** [comment]
git version 2.14.3.windows/1 
same issue

**Comment by parkovski** [comment]
Right now I’m hoping that OpenSSH fixes this - https://github.com/PowerShell/Win32-OpenSSH/issues/914 - and that whatever they do also works for winpty, because I spent a lot of time disassembling conhost and cmd and still couldn’t get it working. Windows generates an SEH exception instead when a debugger is attached so it’s really hard to tell what would trigger it without a debugger. I’ve tried stepping into kernel32 and undoing the debugger check too and still no luck.

My current solution is to patch node-pty’s winpty DLL with one that sends WM_KEYDOWN for Ctrl-C, which works for VSCode on Win10 but is not a general solution for winpty.

WinRS supposedly supports this if anyone is up for the disassembly challenge, but other than that I’m out of ideas.

**Comment by xxxZeus** [comment]
You can end the process using the command `killtask /F /IM node.exe`, for convenience just create a .bat with the above command and save it on the desktop and run it whenever you need to stop the process.

**Comment by parkovski** [comment]
Right, the problem is taskkill tells the kernel to stop the process. If a program catches SIGINT or CTRL_C_EVENT, they can do cleanup first, or sometimes (shells for example) it’s not even appropriate for them to quit. The core problem here is while it’s hard but not impossible to send that to an arbitrary process on Windows, cmd.exe seems to ignore it unless it comes from conhost, which makes it hard to write a decent terminal emulator.

**Comment by Tyriar** [comment]
Weird things happen when exiting `git log` via ctrl+c as well, see https://github.com/Microsoft/vscode/issues/39419

**Comment by parkovski** [comment]
That's reproducible in the regular console using [this utility](https://github.com/parkovski/isvt) - turn off processed input using `isvt -p (pid of shell process running git log) i=_-ENABLE_PROCESSED_INPUT`, then type `^Z ^C`. The git process exits but it's leaving behind the less process it spawned, although I'm not sure exactly which part of that sequence of events causes it.

**Comment by cScarlson** [comment]
Same. Child process doesn't close. I switch my server's port a few times, and then have to manually search & destroy the processes -- or only lift my server using Command Prompt. I've seen older git-issues of this problem. Just update Git from 2.6.x to 2.15.1 assuming this was critical enough to be fixed by now.

`git --version`: `2.15.1.windows.2` (Windows 10 | 64)
`node -v`: `v8.9.3`

**Comment by sedulous-mortal** [comment]
I was able to get all my node processes to die directly from the Git Bash shell on Windows 10 by typing `taskkill -F -IM node.exe` - this ends all the node processes on my computer at once. I found I could also use `taskkill //F //IM node.exe`. Not sure why both `-` and `//` work in this context. Hope this helps!

**Comment by parkovski** [comment]
I've confirmed (by tracing API calls) that this stems from a bug in cmd.exe - https://github.com/rprichard/winpty/issues/116#issuecomment-352837629

Winpty's handling doesn't cover unprocessed mode last time I checked, but I think that's just from the confusion around cmd. Hoping there's a way to set the keyboard state for cmd, otherwise it'd have to be fixed by MS or hook an API call.

**Comment by florian-kittel** [comment]
I have the same issue. I tried it today with the latest git update to 2.16.0. 

It now cancels my node js (server and angular cli) applications in a git bash, but I have still the issue at an angular application running in a bash terminal at VS code. It takes 30 seconds or more to cancel.

**Comment by janacm** [comment]
Quick solution for NodeJS issues: use powershell as an external terminal

**Comment by SharakPL** [comment]
Don't know about NodeJS but exitig `git log` with `ctrl+c` bugs the terminal completely. It's not VSCode related or even console related because the same thing happens for me in Git-bash (mingw64), Powershell and CMD in windows, even outside VSCode. It must be one of Git's dependency that is also used by Node.

**Comment by Mellbourn** [comment]
I think I have the same issue in tmux terminal inside VSCode on MacOS.

I have the setup that was described by João Moreno (@joaomoreno) in his blog post https://medium.com/@joaomoreno/persistent-terminal-sessions-in-vs-code-8fc469ed6b41

However, it seems to me that Control-C, does not work as good as it used to.

It is particularly noticeable when I run `nodemon`. When I run it in tmux in `iTerm`, I can exit the run without problem by using Control-C. If I run `nodemon` inside tmux inside VSCode, I cannot. Instead it just renders `^C` in the terminal window.


- VSCode Version: 1.21.1 and 1.22.1-insiders b75321329388d74e34f79dcfcb789246e31836a8
- OS Version: 10.13.3


Does this issue occur when all extensions are disabled?: Yes

**Comment by Tyriar** [comment]
@Mellbourn that's definitely not this issue as this is Windows only, we can follow up in a VS Code issue if you want to create one https://github.com/Microsoft/vscode

**Comment by Mellbourn** [comment]
@Tyriar Right. I've reported the problem here now: https://github.com/Microsoft/vscode/issues/46731

**Comment by dieselVtwin** [comment]
Hi all,
I just bumped into this issue, but without VS Code involved in the mix.

If this is only an issue on Windows platforms, as @Tyriar states, then I think it is actually under VS Code and in the Bash shell included in GIT for Windows.  It's discussed in issue 1491.
https://github.com/git-for-windows/git/issues/1491

A Linux engineer that I worked with helped this poor Windows engineer to work around it with the magical phrase, "stty echo", to turn the keyboard echo back on.

Cheers,
dVt

**Comment by Abdurraheem** [comment]
I had the same issue.
GitBash - git version 2.16.2.windows.1

When I use e.g. git log command and press ctrl+c the console is stuck. I could type, but no command are typing. However, after press enter the command is running.But I got solution ,after git log , press wq to exit from the log and now my console cursor is visible and and command is visible now.

**Comment by parkovski** [comment]
Hey @Abdurraheem I believe that’s actually a git bug, try killing all processes named less after you get out of the git log.

**Comment by haxor211** [comment]
So it this ever gonna get fixed? Or Im doomed to use taskkill /im node.exe /f forever?

**Comment by Tyriar** [comment]
@haxor211 I'm hoping the fix for https://github.com/Microsoft/vscode/issues/45693 will do the trick, I'll be working on that probably in the next few months.

**Comment by saostad** [comment]
I have same problem with git bash, but after press ctrl+c, if I wait like 5 minutes it will stop the "npm start" process for me. 
I am using create-react-app typescript package
but if I use "docker-compose up"  to run my docker container after I press ctrl+c it stop the container without problem.

**Comment by saostad** [comment]
I find what was my problem.
for some reason I had this in my settings:
`"terminal.integrated.rendererType": "dom"`
after I remove it it start working and it's much faster

**Comment by pikilon** [comment]
>"terminal.integrated.rendererType": "dom"
In my case I put this to "dom" and problem solved

**Comment by gpmetheny** [comment]
> I had the same issue.
> GitBash - git version 2.16.2.windows.1
> 
> When I use e.g. git log command and press ctrl+c the console is stuck. I could type, but no command are typing. However, after press enter the command is running.But I got solution ,after git log , press wq to exit from the log and now my console cursor is visible and and command is also visible now.

I was having the same problem specifically with `git log`, but that's because it uses `less`. Using `q` (NOT ctrl + q) should exit properly and allow you to continue typing commands as normal.

\*\*\* *answer found [here](https://stackoverflow.com/questions/9483757/how-to-exit-git-log-or-git-diff#answer-9483803)* \*\*\*

**Comment by NischalSehrawat** [comment]
CTRL+ALT+M

**Comment by luohong123** [comment]
I am experiencing the same problem, when I press CTRL+C, the editor unexpectedly quits, appears 3221225786, my solution is to upgrade the git tool, the following is my git version
```bash
git --version
// git version 2.23.0.windows.1
```
The issue has been resolved after the upgrade

**Comment by tohagan** [comment]
If you start another bash shell inside the bash started by VSCode then I found that apps you start in the inner shell DO respond to Ctrl-C.

**Comment by rileyhemp** [comment]
Running git update-git-for-windows did the trick

**Comment by FiruzzZ** [comment]
I have this issue with 2.25.1 (just updated from 2.25.0) to check if it was already fixed but no.
CTRL+C on Gitbash works with  commands like:
` ping xxx.xxx.xxx //t`
but doesn't work with git log or git diff, I have to press ENTER a few time after pressing CTRL+C and even after this, it takes a few seconds to the interrupt the command

**Comment by Tyriar** [comment]
@FiruzzZ fyi in bash, git log/diff aren't meant to exit with ctrl+C,.

**Comment by setebe** [comment]
The temporary solution for me: 
 
![image](https://user-images.githubusercontent.com/26058408/75633262-3f135b80-5c25-11ea-95ed-2de08edf0079.png)

**Comment by SharakPL** [comment]
@setebe I think much better solution is to simply remember to use `Q` instead of `Ctrl C` in scrollable parts like _git log_ output. Besides I've just checked and using `Ctrl C` while viewing _git log_ does nothing in powershell and cmd. You must use `Q` to quit log so **this is already fixed**.

I'm using _Git for Windows 2.25.1.windows.1 (64bit)_

**Comment by setebe** [comment]
@SharakPL I've downgraded to version 2.8.1 and ctrl+c starts to work.

**Comment by Tyriar** [comment]
ctrl+c is not meant to work in git log/diff, downgrading is not recommended as you'll probably hit other bugs that have been fixed in it. Anyway this issue is long fixed with the move to conpty so this discussion is off topic.
